File: cpp/Heap_PriorityQueue/295.find-median-from-data-stream.cpp
#include "../std.hpp"
class MedianFinder {
    priority_queue<int> lower;
    priority_queue<int, vector<int>, greater<int>> upper;

public:
    MedianFinder()
    {
    }
    void addNum(int num)
    {
        lower.push(num);
        if (lower.size() > upper.size() + 1 || !upper.size() == 0 && lower.top() > upper.top()) {
            upper.push(lower.top());
            lower.pop();
        }
        if (upper.size() > lower.size() + 1) {
            lower.push(upper.top());
            upper.pop();
        }
    }

    double findMedian()
    {
        int maxSize = lower.size();
        int minSize = upper.size();
        if (maxSize == minSize) {
            return (double)(upper.top() + lower.top()) / 2;
        }
        else if (maxSize > minSize) {
            return (double)lower.top();
        }
        else {
            return (double)upper.top();
        }
    }
};
int main()
{
    MedianFinder *obj = new MedianFinder();
    obj->addNum(1);
    obj->addNum(2);
    double param_2 = obj->findMedian();
}

File: TypeScript/Heap_PriorityQueue/295.find-median-from-data-stream.ts
import {
  MinPriorityQueue,
  MaxPriorityQueue,
} from "@datastructures-js/priority-queue"

class MedianFinder {
  lower = new MaxPriorityQueue({ priority: num => num })
  upper = new MinPriorityQueue({ priority: num => num })
  constructor() {}

  addNum(num: number): void {
    this.lower.enqueue(num)
    if (
      this.lower.size() > this.upper.size() + 1 ||
      (this.upper.size() != 0 && this.lower.front().element > this.upper.front().element)
    ) {
      this.upper.enqueue(this.lower.front().element)
      this.lower.dequeue()
    }
    if (this.upper.size() > this.lower.size() + 1) {
      this.lower.enqueue(this.upper.front().element)
      this.upper.dequeue()
    }
  }

  findMedian(): number {
    let maxSize = this.lower.size()
    let minSize = this.upper.size()
    if (maxSize == minSize) {
      return (this.upper.front().element + this.lower.front().element) / 2
    } else if (maxSize > minSize) {
      return this.lower.front().element
    } else {
      return this.upper.front().element
    }
  }
}
let mf = new MedianFinder()
mf.addNum(1)
console.log(mf.findMedian())
